docs/api/adbc.md

```diff
@@ -123,3 +130,82 @@ StatementSetOption(&adbc_statement, ADBC_INGEST_OPTION_TARGET_TABLE, "AnswerToEv
 StatementBindStream(&adbc_statement, &arrow_stream, &adbc_error);
 StatementExecuteQuery(&adbc_statement, nullptr, nullptr, &adbc_error);
 ```
+
+### Python
+
+The first thing to do is to use `pip` and install the ADBC Driver manager.
+
+```shell
+pip install adbc_driver_manager
+```
+
+You will also need the `pyarrow` package to directly access Apache Arrow
+formatted result sets.
+```shell
+pip install pyarrow
+```
+
+The full documentation for the `adbc_driver_manager` package can be found
+[here](https://arrow.apache.org/adbc/current/python/api/adbc_driver_manager.html).
+
+As with C++, we need to provide initialization options 
+```python
+import adbc_driver_manager.dbapi
+conn = adbc_driver_manager.dbapi.connect(
+    driver="path/to/libduckdb.dylib",
+    entrypoint="duckdb_adbc_init",
+    db_kwargs={"path": "test.db"},
+)
+
+with conn.cursor() as cur:
+    pass
+
+conn.close()
+```
+
+It is also possible to leverage the duckdb Python library directly:
+```python
+import duckdb
+import adbc_driver_manager.dbapi
+conn = adbc_driver_manager.dbapi.connect(
+    driver=duckdb.__file__,
+    entrypoint="duckdb_adbc_init",
+    db_kwargs={"path": "test.db"}
+)
+
+with conn.cursor() as cur:
+    pass
+
+conn.close()
+```
+
+Now that we have a connection initialized, we can initialize a cursor
+for running queries. 
+
+```python
+import pyarrow
+...
+
+with conn.cursor() as cur:
+    cur.execute("SELECT 42")
+    # fetch as a pyarrow table
+    tbl = cur.fetch_arrow_table()
+    # alternately as a pandas dataframe
+    df = cur.fetch_df()
+    # you can also use fetchone/fetchall
+    # and fetchmany as per DBAPI (PEP 249)
+```
+
+Data can also be ingested via `arrow_streams`. We just need to set options
```

Comment:
```suggestion
Data can also be ingested via `adbc_ingest`. We just need to set options
```

Did you mean this?

Reply:
Hmm i was following the documentation prose from the C++ example. I think it was explicitly saying you can use Arrow data record batches to ingest into the db.

Reply:
If I run:
```py
import duckdb
import adbc_driver_manager.dbapi
import pyarrow

conn = adbc_driver_manager.dbapi.connect(
    driver=duckdb.__file__,
    entrypoint="duckdb_adbc_init",
    db_kwargs={"path": "test.db"}
)

data = pyarrow.record_batch(
    [[1, 2, 3, 4], ["a", "b", "c", "d"]],
    names=["ints", "strs"],
)

with conn.cursor() as cur:
    cur.adbc_ingest("AnswerToEverything", data)   
```

I get:
```text
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/opt/homebrew/lib/python3.11/site-packages/adbc_driver_manager/dbapi.py", line 769, in adbc_ingest
    self._stmt.set_options(
  File "adbc_driver_manager/_lib.pyx", line 1114, in adbc_driver_manager._lib.AdbcStatement.set_options
  File "adbc_driver_manager/_lib.pyx", line 385, in adbc_driver_manager._lib.check_error
adbc_driver_manager._lib.ProgrammingError: ADBC_STATUS_INVALID_ARGUMENT (5)
```

Do you have any suggestion why this happens?

Reply:
> 

This needs `duckdb.duckdb.__file__`, but this was written at a time where we did not have `adbc_driver_duckdb` yet, this logic has been wrapped up in that module.

See https://github.com/duckdb/duckdb/blob/main/tools/pythonpkg/tests/fast/adbc/test_statement_bind.py for an example of how to work with the module

Reply:
@Tishj If i try `duckdb.duckdb.__file__` I get 

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'duckdb' has no attribute 'duckdb'
```

It also appears that it's duckdb that is returning `ADBC_STATUS_INVALID_ARGUMENT` as an error when calling `AdbcStatementSetOption` with key `adbc.ingest.mode` and value `adbc.ingest.mode.create`, which is internally called by `adbc_ingest`. 

I'm using `python-duckdb` v0.8.1, is there something I'm missing? The test file you linked doesn't use `adbc_ingest` at all, nor does it try calling `set_option` on the statement.

Reply:
Ah, i figured out the issue: v0.8.1 of duckdb has the following:

![image](https://github.com/duckdb/duckdb-web/assets/555095/0b0e5ae9-8418-4ea2-bd54-e3cf9919d542)

The proper handling of the ingest options was only committed *after* the v0.8.1 cut and so support for `adbc_ingest` currently only exists if you build off main, not if you're using v0.8.1.

Should we hold off on this until the next version is cut? Or should I just remove this section for now and re-add it with the next version of duckdb?

Reply:
The docs are targeting the next release, so it shouldbe added to the docs before 0.9 is released :+1:

Reply:
@szarnyasg Can you try the code snippet with a 0.9 build of duckdb and confirm that it works for you too?

Reply:
@zeroshade the documentation for v0.8.1 is already archived, so this PR is targeting the `dev` version (which will turn into v0.9.0 later this month). So let's get this merged!

I ran the code again on a bleeding edge DuckDB package:
```py
>>> import duckdb
>>> print(duckdb.__version__)
0.8.2-dev3250
```

I got a different error now:

```text
>>> conn = adbc_driver_manager.dbapi.connect(
...     driver=duckdb.__file__,
...     entrypoint="duckdb_adbc_init",
...     db_kwargs={"path": "test.db"}
... )
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/lib/python3.11/site-packages/adbc_driver_manager/dbapi.py", line 200, in connect
    db = _lib.AdbcDatabase(**db_kwargs)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "adbc_driver_manager/_lib.pyx", line 569, in adbc_driver_manager._lib.AdbcDatabase.__init__
  File "adbc_driver_manager/_lib.pyx", line 399, in adbc_driver_manager._lib.check_error
adbc_driver_manager.InternalError: INTERNAL: dlopen() failed: dlopen(/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py, 0x0006): tried: '/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py' (not a mach-o file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py' (no such file), '/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py' (not a mach-o file)
dlopen() failed: dlopen(lib/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py.dylib, 0x0006): tried: 'lib/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlib/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py.dylib' (no such file), '/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/Resources/Python.app/Contents/MacOS/../../../../../../../../../../../lib/lib/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py.dylib' (no such file), '/usr/lib/lib/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py.dylib' (no such file, not in dyld cache), 'lib/opt/homebrew/lib/python3.11/site-packages/duckdb/__init__.py.dylib' (no such file)
```

I'm uncertain about this - is this a problem with my setup? Please advise.

Reply:
It looks like for some reason the value of `duckdb.__file__` for you was pointing to `.../duckdb/__init__.py` rather than to the `libduckdb.dylib` correctly. Maybe something about how you built it?

Reply:
As said before, it's no longer `duckdb.__file__`, it's `duckdb.duckdb.__file__` and this is abstracted away with `adbc_driver_duckdb`

Also the entrypoint symbol is not exported when built with `BUILD_PYTHON=1 make`
Only through `python3 setup.py install` (run from the `tools/pythonpkg` folder)

Reply:
I downloaded and installed the bleeding edge version myself and got it to work, i'll update the docs here.

Reply:
@Tishj updated, let me know what you think.

@szarnyasg Try the new example with the bleeding edge build?

Reply:
> @szarnyasg Try the new example with the bleeding edge build?

It works fine now, thanks!

Reply:
Thank you, looks good to me!

I really want to discourage using the raw connect logic, it's needlessly complex and as seen in 0.8.1 -> dev; easily broken

So thanks for using the abstraction!

Reply:
Great! @zeroshade do you have any further changes to commit or is this good to be merged?

Reply:
This is good to be merged! I'm gonna file other PRs for adding docs for other languages to this. :smile:

Reply:
Great, thanks!